experimental/cuda2/graph_command_buffer.c

```diff
@@ -26,6 +25,154 @@
 // descriptor bindings and push constants.
 #define IREE_HAL_CUDA_MAX_KERNEL_ARG 128
 
+// Segmented submission management
+//
+// In a CUDA graph, buffer management and kernel launches are represented as
+// graph nodes. Dependencies are represented by graph edges. IREE's HAL follows
+// the Vulkan command buffer recording model, which "linearizes" the original
+// graph. So we have a mismatch here. Implementing IREE's HAL using CUDA graph
+// would require rediscover the graph node dependencies from the linear chain of
+// command buffer commands; it means looking at both previous and next commands
+// sometimes.
+//
+// Due to these reasons, it's beneficial to have a complete view of the full
+// command buffer and extra flexibility during recording, in order to fixup past
+// commands, or inspect past/future commands.
+//
+// Therefore, to implement IREE HAL command buffers using CUDA graphs, we
+// perform two steps using a linked list of command segments. First we create
+// segments (iree_hal_cuda2_command_buffer_prepare_*) to keep track of all IREE
+// HAL commands and the associated data, and then, when finalizing the command
+// buffer, we iterate through all the segments and record their contents
+// (iree_hal_cuda2_command_segment_record_*) into a proper CUDA graph command
+// buffer. A linked list gives us the flexibility to organize command sequence
+// in low overhead; and a deferred recording gives us the complete picture of
+// the command buffer when really started recording.
```

Comment:
How much overhead is this logic adding? Would it be possible to do some of this ahead of time in the compiler and pass extra data through to the executable flatbuffer?

Reply:
> How much overhead is this logic adding?

It should be minor overhead at command buffer composition time due to the linked list--we need to create the list and scan it once/twice later when emitting the concrete commands. It takes extra allocations and time, but both at O(N) where N is the number of commands. We'd need the flexibility here to handle barriers and such for example in https://github.com/openxla/iree/pull/14526. And we can go even fancier later. 

> Would it be possible to do some of this ahead of time in the compiler and pass extra data through to the executable flatbuffer?

Not really. Fundamentally this is due to the semantics mismatch between CUDA graph and  IREE HAL. Unless we want to change how HAL are defined.